File: scripts/conventional_commits_test.py
"""Tests for the shared Conventional Commit classification.

Standard library unittest, because this is the only Python in the repository
that needs testing and a test runner is a dependency to install, pin and keep
current for one file. Run with `task quality:scripts:test`.

What matters here is the boundary: which commits move the version and which
count as breaking. Both are read by the release workflow, which cuts versions
from them, and by the release-flow gate, which refuses breaking changes on a
pull request into main. A disagreement between the two is the failure the shared
module exists to make impossible, so these cases are the contract.
"""

import unittest

import conventional_commits as cc


def commits(*subjects_and_bodies):
    """Build commits from (subject, body) pairs, or bare subjects."""
    built = []
    for index, item in enumerate(subjects_and_bodies):
        subject, body = item if isinstance(item, tuple) else (item, "")
        built.append(cc.classify(f"{index:040x}", subject, body))

    return built


class ClassifySubject(unittest.TestCase):
    def test_reads_type_scope_and_description(self):
        commit = cc.classify("abc", "feat(auth): add a thing", "")
        self.assertTrue(commit.conventional)
        self.assertEqual(commit.type, "feat")
        self.assertEqual(commit.scope, "auth")
        self.assertEqual(commit.description, "add a thing")
        self.assertFalse(commit.breaking)

    def test_scope_is_optional(self):
        commit = cc.classify("abc", "fix: a thing", "")
        self.assertEqual(commit.type, "fix")
        self.assertIsNone(commit.scope)

    def test_bang_marks_breaking_with_and_without_scope(self):
        self.assertTrue(cc.classify("a", "feat!: a thing", "").breaking)
        self.assertTrue(cc.classify("a", "feat(api)!: a thing", "").breaking)

    def test_unconventional_subject_is_other_and_never_conventional(self):
        commit = cc.classify("abc", "Revert to a project per test", "")
        self.assertFalse(commit.conventional)
        self.assertEqual(commit.type, "other")
        self.assertEqual(commit.description, "Revert to a project per test")


class BreakingFooter(unittest.TestCase):
    def test_both_spellings_count(self):
        self.assertTrue(cc.classify("a", "fix: x", "BREAKING CHANGE: gone").breaking)
        self.assertTrue(cc.classify("a", "fix: x", "BREAKING-CHANGE: gone").breaking)

    def test_footer_may_follow_a_body(self):
        body = "Some prose about the change.\n\nBREAKING CHANGE: the flag is gone"
        self.assertTrue(cc.classify("a", "fix: x", body).breaking)

    def test_mid_line_mention_is_not_a_footer(self):
        # The reason this module exists. Bump detection used to ask whether the
        # body contained the string anywhere, so a sentence mentioning it cut a
        # major release while the changelog -- which matched a footer anchored
        # to a line -- did not list it as breaking. A footer is what the
        # specification describes, and it is what the notes have reported.
        body = "This is deliberately not a BREAKING CHANGE: it only looks like one."
        self.assertFalse(cc.classify("a", "fix: x", body).breaking)

    def test_an_unconventional_subject_still_carries_a_footer(self):
        commit = cc.classify("a", "Merge branch whatever", "BREAKING CHANGE: gone")
        self.assertFalse(commit.conventional)
        self.assertTrue(commit.breaking)


class BumpLevel(unittest.TestCase):
    def test_nothing_releasing_is_none(self):
        self.assertEqual(
            cc.bump_level(commits("chore: x", "docs: y", "test: z", "refactor: w")),
            cc.BUMP_NONE,
        )

    def test_fix_perf_and_revert_are_patch(self):
        for subject in ("fix: x", "perf: x", "revert: x"):
            with self.subTest(subject=subject):
                self.assertEqual(cc.bump_level(commits(subject)), cc.BUMP_PATCH)

    def test_feat_is_minor(self):
        self.assertEqual(cc.bump_level(commits("feat: x")), cc.BUMP_MINOR)

    def test_breaking_is_major_however_it_is_spelled(self):
        self.assertEqual(cc.bump_level(commits("feat!: x")), cc.BUMP_MAJOR)
        self.assertEqual(
            cc.bump_level(commits(("fix: x", "BREAKING CHANGE: gone"))),
            cc.BUMP_MAJOR,
        )

    def test_the_highest_wins(self):
        self.assertEqual(
            cc.bump_level(commits("chore: a", "fix: b", "feat: c")),
            cc.BUMP_MINOR,
        )
        self.assertEqual(
            cc.bump_level(commits("chore: a", "feat: b", "fix!: c")),
            cc.BUMP_MAJOR,
        )

    def test_an_unconventional_subject_does_not_release(self):
        # Not even with a breaking footer: a subject the parser cannot read has
        # no type to release under, and the non-conventional revert commit on
        # the v4 branch is exactly this shape.
        self.assertEqual(
            cc.bump_level(commits(("Revert something", "BREAKING CHANGE: gone"))),
            cc.BUMP_NONE,
        )


class ParseLog(unittest.TestCase):
    def test_reads_the_log_format_it_declares(self):
        raw = (
            "1111111\x1ffeat(x): first\x1f\x1e"
            "2222222\x1ffix: second\x1fBREAKING CHANGE: gone\x1e"
        )
        parsed = cc.parse_log(raw)
        self.assertEqual([c.sha for c in parsed], ["1111111", "2222222"])
        self.assertEqual(parsed[0].type, "feat")
        self.assertTrue(parsed[1].breaking)

    def test_ignores_empty_records(self):
        self.assertEqual(cc.parse_log(""), [])
        self.assertEqual(cc.parse_log("\x1e\x1e"), [])

    def test_has_conventional_distinguishes_none_from_non_releasing(self):
        self.assertFalse(cc.has_conventional(commits("Merge pull request #1")))
        self.assertTrue(cc.has_conventional(commits("chore: x")))


if __name__ == "__main__":
    unittest.main()

File: tools/conventionalcommits/conventionalcommits_test.go
package conventionalcommits

import (
	"fmt"
	"testing"
)

// buildCommits makes commits from (subject, body) pairs so a bump case reads as
// the list of commits it is about.
func buildCommits(pairs ...[2]string) []Commit {
	built := make([]Commit, 0, len(pairs))
	for index, pair := range pairs {
		built = append(built, Classify(fmt.Sprintf("%040x", index), pair[0], pair[1]))
	}

	return built
}

func subjects(list ...string) [][2]string {
	pairs := make([][2]string, 0, len(list))
	for _, subject := range list {
		pairs = append(pairs, [2]string{subject, ""})
	}

	return pairs
}

func TestClassifyReadsTypeScopeAndDescription(t *testing.T) {
	t.Parallel()

	commit := Classify("abc", "feat(auth): add a thing", "")
	if !commit.Conventional {
		t.Fatalf("expected a conventional commit, got %+v", commit)
	}
	if commit.Type != "feat" {
		t.Errorf("type: got %q, want feat", commit.Type)
	}
	if commit.Scope == nil || *commit.Scope != "auth" {
		t.Errorf("scope: got %v, want auth", commit.Scope)
	}
	if commit.Description != "add a thing" {
		t.Errorf("description: got %q, want %q", commit.Description, "add a thing")
	}
	if commit.Breaking {
		t.Error("expected not breaking")
	}
}

func TestClassifyScopeIsOptional(t *testing.T) {
	t.Parallel()

	commit := Classify("abc", "fix: a thing", "")
	if commit.Type != "fix" {
		t.Errorf("type: got %q, want fix", commit.Type)
	}
	if commit.Scope != nil {
		t.Errorf("scope: got %q, want none", *commit.Scope)
	}
}

func TestClassifyBangMarksBreakingWithAndWithoutScope(t *testing.T) {
	t.Parallel()

	for _, subject := range []string{"feat!: a thing", "feat(api)!: a thing"} {
		if !Classify("a", subject, "").Breaking {
			t.Errorf("%q: expected breaking", subject)
		}
	}
}

func TestClassifyUnconventionalSubjectIsOtherAndNeverConventional(t *testing.T) {
	t.Parallel()

	commit := Classify("abc", "Revert to a project per test", "")
	if commit.Conventional {
		t.Error("expected not conventional")
	}
	if commit.Type != "other" {
		t.Errorf("type: got %q, want other", commit.Type)
	}
	if commit.Description != "Revert to a project per test" {
		t.Errorf("description: got %q", commit.Description)
	}
}

func TestBreakingFooterBothSpellingsCount(t *testing.T) {
	t.Parallel()

	for _, body := range []string{"BREAKING CHANGE: gone", "BREAKING-CHANGE: gone"} {
		if !Classify("a", "fix: x", body).Breaking {
			t.Errorf("%q: expected breaking", body)
		}
	}
}

func TestBreakingFooterMayFollowABody(t *testing.T) {
	t.Parallel()

	body := "Some prose about the change.\n\nBREAKING CHANGE: the flag is gone"
	if !Classify("a", "fix: x", body).Breaking {
		t.Error("expected breaking")
	}
}

// The reason this package exists. Bump detection used to ask whether the body
// contained the string anywhere, so a sentence mentioning it cut a major
// release while the changelog -- which matched a footer anchored to a line --
// did not list it as breaking. A footer is what the specification describes,
// and it is what the notes have reported.
func TestBreakingFooterMidLineMentionIsNotAFooter(t *testing.T) {
	t.Parallel()

	body := "This is deliberately not a BREAKING CHANGE: it only looks like one."
	if Classify("a", "fix: x", body).Breaking {
		t.Error("expected a mid-line mention not to count")
	}
}

func TestBreakingFooterAnUnconventionalSubjectStillCarriesAFooter(t *testing.T) {
	t.Parallel()

	commit := Classify("a", "Merge branch whatever", "BREAKING CHANGE: gone")
	if commit.Conventional {
		t.Error("expected not conventional")
	}
	if !commit.Breaking {
		t.Error("expected breaking")
	}
}

// A footer wrapped at the usual commit width has to arrive whole. Reading only
// its first line is what once truncated five of six breaking changes in a
// release, one of them mid-word, and those notes are the migration
// instructions.
func TestBreakingFooterIsCollectedToTheBlankLineAndCollapsed(t *testing.T) {
	t.Parallel()

	body := "Some prose.\n\n" +
		"BREAKING CHANGE: --token is gone because a flag value reaches the\n" +
		"process table. Read the token from stdin instead.\n" +
		"\n" +
		"Refs: #123"

	commit := Classify("a", "feat: x", body)
	want := "--token is gone because a flag value reaches the process table. Read the token from stdin instead."
	if commit.BreakingNote != want {
		t.Errorf("note:\n got %q\nwant %q", commit.BreakingNote, want)
	}
}

func TestBreakingFooterNoteIsEmptyWithoutOne(t *testing.T) {
	t.Parallel()

	if note := Classify("a", "feat!: x", "").BreakingNote; note != "" {
		t.Errorf("note: got %q, want empty", note)
	}
}

func TestBumpLevelNothingReleasingIsNone(t *testing.T) {
	t.Parallel()

	got := BumpLevel(buildCommits(subjects("chore: x", "docs: y", "test: z", "refactor: w")...))
	if got != BumpNone {
		t.Errorf("got %v, want BumpNone", got)
	}
}

func TestBumpLevelFixPerfAndRevertArePatch(t *testing.T) {
	t.Parallel()

	for _, subject := range []string{"fix: x", "perf: x", "revert: x"} {
		if got := BumpLevel(buildCommits([2]string{subject, ""})); got != BumpPatch {
			t.Errorf("%q: got %v, want BumpPatch", subject, got)
		}
	}
}

func TestBumpLevelFeatIsMinor(t *testing.T) {
	t.Parallel()

	if got := BumpLevel(buildCommits([2]string{"feat: x", ""})); got != BumpMinor {
		t.Errorf("got %v, want BumpMinor", got)
	}
}

func TestBumpLevelBreakingIsMajorHoweverItIsSpelled(t *testing.T) {
	t.Parallel()

	if got := BumpLevel(buildCommits([2]string{"feat!: x", ""})); got != BumpMajor {
		t.Errorf("bang: got %v, want BumpMajor", got)
	}
	if got := BumpLevel(buildCommits([2]string{"fix: x", "BREAKING CHANGE: gone"})); got != BumpMajor {
		t.Errorf("footer: got %v, want BumpMajor", got)
	}
}

func TestBumpLevelTheHighestWins(t *testing.T) {
	t.Parallel()

	if got := BumpLevel(buildCommits(subjects("chore: a", "fix: b", "feat: c")...)); got != BumpMinor {
		t.Errorf("got %v, want BumpMinor", got)
	}
	if got := BumpLevel(buildCommits(subjects("chore: a", "feat: b", "fix!: c")...)); got != BumpMajor {
		t.Errorf("got %v, want BumpMajor", got)
	}
}

// Not even with a breaking footer: a subject the parser cannot read has no type
// to release under, and the non-conventional revert commit on the v4 branch is
// exactly this shape.
func TestBumpLevelAnUnconventionalSubjectDoesNotRelease(t *testing.T) {
	t.Parallel()

	got := BumpLevel(buildCommits([2]string{"Revert something", "BREAKING CHANGE: gone"}))
	if got != BumpNone {
		t.Errorf("got %v, want BumpNone", got)
	}
}

func TestParseLogReadsTheLogFormatItDeclares(t *testing.T) {
	t.Parallel()

	raw := "1111111\x1ffeat(x): first\x1f\x1e" +
		"2222222\x1ffix: second\x1fBREAKING CHANGE: gone\x1e"

	parsed := ParseLog(raw)
	if len(parsed) != 2 {
		t.Fatalf("got %d commits, want 2", len(parsed))
	}
	if parsed[0].SHA != "1111111" || parsed[1].SHA != "2222222" {
		t.Errorf("shas: got %q and %q", parsed[0].SHA, parsed[1].SHA)
	}
	if parsed[0].Type != "feat" {
		t.Errorf("type: got %q, want feat", parsed[0].Type)
	}
	if !parsed[1].Breaking {
		t.Error("expected the second commit to be breaking")
	}
}

// git writes a newline between records, so every record after the first arrives
// with one in front of its sha.
func TestParseLogToleratesTheNewlineGitPutsBetweenRecords(t *testing.T) {
	t.Parallel()

	raw := "1111111\x1ffeat: first\x1f\x1e\n2222222\x1ffix: second\x1f\x1e"

	parsed := ParseLog(raw)
	if len(parsed) != 2 {
		t.Fatalf("got %d commits, want 2", len(parsed))
	}
	if parsed[1].SHA != "2222222" {
		t.Errorf("sha: got %q, want 2222222", parsed[1].SHA)
	}
}

func TestParseLogIgnoresEmptyRecords(t *testing.T) {
	t.Parallel()

	if got := ParseLog(""); len(got) != 0 {
		t.Errorf("empty: got %d commits, want 0", len(got))
	}
	if got := ParseLog("\x1e\x1e"); len(got) != 0 {
		t.Errorf("separators only: got %d commits, want 0", len(got))
	}
}

func TestHasConventionalDistinguishesNoneFromNonReleasing(t *testing.T) {
	t.Parallel()

	if HasConventional(buildCommits([2]string{"Merge pull request #1", ""})) {
		t.Error("expected an unconventional subject not to count")
	}
	if !HasConventional(buildCommits([2]string{"chore: x", ""})) {
		t.Error("expected chore to count as conventional")
	}
}

func TestShortSHAAbbreviatesAndToleratesShortInput(t *testing.T) {
	t.Parallel()

	if got := (Commit{SHA: "0123456789abcdef"}).ShortSHA(); got != "0123456" {
		t.Errorf("got %q, want 0123456", got)
	}
	if got := (Commit{SHA: "abc"}).ShortSHA(); got != "abc" {
		t.Errorf("got %q, want abc", got)
	}
}
